## Importing libreries

---

In [6]:
# Importing libraries

# Data treatment
# -----------------------------------------------------------------------
import pandas as pd
import numpy as np

# Path
import sys
sys.path.append('../')

# Config
# -----------------------------------------------------------------------
pd.set_option('display.max_columns', None)

from src.support import find_outliers_iqr

import warnings
warnings.filterwarnings('ignore')

## Load data

---

In [7]:
path = "../data/output/data_clean.csv"

df = pd.read_csv(path)
# Conveting dates to datetime
df['Posting Date'] = pd.to_datetime(df['Posting Date'], dayfirst=True)

#### Currency

Although data does not explicitly mention the currency, since these are Brazilian government data, it is reasonable to assume that the amounts are in Brazilian Reais (BRL), as this is the official currency of Brazil. This would be consistent with the standard practices for publishing financial data by the government of that country.

## Revenue Distribution by Economic Category

---

Let's start analyzing the most significant revenue categories and their contribution to total revenues. We can check quickly the different types of Economic Categories.

In [9]:
df['Economic Category'].unique()

array(['Current Revenues', 'Capital Revenues', 'No Information',
       'Intra-Budgetary Current Revenues',
       'Intra-Budgetary Capital Revenues'], dtype=object)

Show the number of operations with an Actual Amount registered grouped by Economic Category

In [10]:
df.groupby('Economic Category')['Actual Amount'].count().round(0)

Economic Category
Capital Revenues                     27065
Current Revenues                    906945
Intra-Budgetary Capital Revenues        86
Intra-Budgetary Current Revenues     14566
No Information                       17893
Name: Actual Amount, dtype: int64

With this information, we can show some initial insights about the data:

1. `Current Revenues` is the most represented category with 906945 entries. `Current Revenues` probably include taxes, fees, and other regular government income.

2. `Intra-Budgetary Current Revenues` has far fewer entries, with a total of 14567. This difference indicates that intra-budgetary revenues, those generated within the government's own budget, represent a much smaller portion of total entries.

3. `Capital Revenues` has 27065 entries, indicating that capital revenues (which might include revenue from asset sales, investments, or financing) are also important but not as common as current revenues. However, this kind of transactions are usually way more expensive than `Current Revenues` as we will see in the next step.

4. `Intra-Budgetary Capital Revenues` with only 86 entries, shows that intra-budgetary capital revenues are unusual. This could imply that there are few events or sources of capital income occurring within the government's own budget.

5. `No Information` has 17893 entries. Depending on how big the mean amount is it could be a big loss of information or couldn't be.

Let's now move to the value of operations with an Actual Amount registered grouped by Economic Category

In [18]:
df.groupby('Economic Category')['Actual Amount'].sum().round(0)

Economic Category
Capital Revenues                    1.200415e+13
Current Revenues                    1.202508e+13
Intra-Budgetary Capital Revenues    2.141127e+10
Intra-Budgetary Current Revenues    2.805787e+11
No Information                      3.271665e+11
Name: Actual Amount, dtype: float64

We can extract again some valuable insights:

1. `Current Revenues`:
   - The total actual revenues amount to 1.202 x 10¹³ (approximately 12.02 trillion Brazilian reais). This confirms that current revenues are a huge source of income for the government, as already suggested by the number of entries in this category.

2. `Intra-Budgetary Current Revenues`:
   - The total actual revenues amount to 2.8058 x 10¹¹ (280.58 billion reais). Although much smaller than `Current Revenues`, it is still a significant amount.

3. `Capital Revenues`:
   - The total actual revenues amount to 1.20 x 10¹³ (approximately 12 trillion reais). Capital revenues are almost equivalent in magnitude to current revenues, indicating that this category also plays a crucial role in government revenue.

4. `Intra-Budgetary Capital Revenues`:
   - The total actual amount is 2.141 x 10¹⁰ (21.41 billion reais). These revenues are a small fraction compared to the total capital revenues.

5. `No Information`:
   - The total amount is 3.27166 x 10¹¹ (327.17 billion reais). Represents a 1.33% out of the total with unknown category. Despite seeming to be low it's a huge amount that should have been labeled.

---

Now we can compute the average difference between projected and actual revenues for each category. This only makes sense for entries that have both numbers available, which is only 8837 entries

In [23]:
df_diff = df[df['Actual Amount'].notna() & df['Updated Budgeted Amount'].notna()][['Economic Category', 'Actual Amount', 'Updated Budgeted Amount']]
df_diff

Economic Category  Actual Amount  Updated Budgeted Amount
20       Capital Revenues      985939.65                 640279.0
21       Capital Revenues      194307.50                 358413.0
22         No Information       68323.99                 111097.0
23       Current Revenues      107548.62                 350042.0
25       Current Revenues     5145837.84                4639104.0
...                   ...            ...                      ...
1024565  Current Revenues        1098.41                 907063.0
1025301  Current Revenues         480.00                 517775.0
1025530  Current Revenues        1483.23                1323033.0
1025680  Current Revenues         135.90                 407371.0
1026029  Current Revenues      349087.72               97926843.0

[8837 rows x 3 columns]

In [24]:
df_diff['Difference'] = df_diff['Actual Amount'] - df_diff['Updated Budgeted Amount']
df_diff['Porcentual_difference'] = df_diff['Difference'] / df_diff['Updated Budgeted Amount'] * 100

df_diff.groupby('Economic Category')['Difference'].mean().round(0)

Economic Category
Capital Revenues                   -2.129977e+10
Current Revenues                   -4.509515e+08
Intra-Budgetary Current Revenues   -1.193434e+08
No Information                     -4.687883e+09
Name: Difference, dtype: float64

It makes more sense to show the porcentual difference:

In [25]:
df_diff.groupby('Economic Category')['Porcentual_difference'].agg(['mean', 'median'])

mean     median
Economic Category                                       
Capital Revenues                  2785.083178 -46.251895
Current Revenues                  9900.145758 -48.355853
Intra-Budgetary Current Revenues   199.616772 -38.589202
No Information                    1613.218577 -31.114211

We can see that mean and median are extremely different. This is usually caused by outliers, let's check

In [27]:
# Number of outliers
find_outliers_iqr(df_diff, 'Porcentual_difference').sort_values(by = 'Porcentual_difference').shape[0]

805

There are 805 outliers, but some of them are very extreme:

In [32]:
# Number of outliers using 3000 times the IQR instead of 1.5 (they exist)
outliers = find_outliers_iqr(df_diff, 'Porcentual_difference', width = 3000).sort_values(by = 'Porcentual_difference', ascending=False).index.tolist()

len(outliers)

7

There are a lot of outliers and some of them extremely far from the median

In [33]:
for i in outliers:

    print(df.iloc[i,11])

NaT
2014-12-31 00:00:00
2013-12-31 00:00:00
2015-12-31 00:00:00
2014-12-31 00:00:00
2013-12-31 00:00:00
2013-12-31 00:00:00


The entries being posted on December 31st likely reflect year-end financial adjustments, where revenues accumulated throughout the year are recognized in one batch. This could result from accounting practices like delayed reporting or systematic year-end postings.

For other dates, there may be other causes like underestimated revenues, unexpected demand for services, large projects/contracts not initially accounted for in the budget or even data entry errors or misclassification.

---

### Temporal analysis

In [34]:
# Now, let's evaluate the temporal trends by calculating the sum of actual income per year.
df.groupby('Fiscal Year')['Actual Amount'].sum()

Fiscal Year
2013    1.663430e+12
2014    2.211514e+12
2015    2.634219e+12
2016    2.787181e+12
2017    2.476073e+12
2018    2.865977e+12
2019    2.880424e+12
2020    3.460501e+12
2021    3.679059e+12
Name: Actual Amount, dtype: float64

We can see a increasement of the incomes over time except for a slight decreasement in 2016-2017 that can be related to Brasil economical crisis in 2014-2017

In [35]:
# Now, let's evaluate the temporal trends by calculating the sum of actual income per year and month.
for i in range(2013, 2022):
    
    print(i, df[df['Fiscal Year'] == i].groupby(df['Posting Date'].dt.month)['Actual Amount'].sum())

2013 Posting Date
12.0    1.617267e+12
Name: Actual Amount, dtype: float64
2014 Posting Date
12.0    2.192906e+12
Name: Actual Amount, dtype: float64
2015 Posting Date
12.0    2.630488e+12
Name: Actual Amount, dtype: float64
2016 Posting Date
1.0     2.445406e+11
2.0     1.443923e+11
3.0     3.234955e+11
4.0     2.215033e+11
5.0     1.662483e+11
6.0     1.752807e+11
7.0     2.347508e+11
8.0     1.788412e+11
9.0     1.913556e+11
10.0    2.877622e+11
11.0    1.512721e+11
12.0    4.413237e+11
Name: Actual Amount, dtype: float64
2017 Posting Date
1.0     3.554494e+11
2.0     1.604895e+11
3.0     1.949693e+11
4.0     1.909670e+11
5.0     1.517739e+11
6.0     1.733170e+11
7.0     1.801296e+11
8.0     2.202914e+11
9.0     2.212834e+11
10.0    2.109477e+11
11.0    1.686580e+11
12.0    1.703508e+11
Name: Actual Amount, dtype: float64
2018 Posting Date
1.0     3.058722e+11
2.0     1.599130e+11
3.0     2.478358e+11
4.0     2.647583e+11
5.0     1.734651e+11
6.0     1.754082e+11
7.0     2.402284e+1

1. **2013-2015**: The data only includes the total amount for December, and these amounts are significantly higher compared to individual months in subsequent years. This suggests that income may have been reported in a lump sum at the end of the year during these early years.

2. **2016 onwards**: The income is reported monthly, providing a clearer picture of income flows. Notable trends include:
   - **2016**: A peak in December with an actual amount of approximately 4.41 x 10¹¹.

   - **2017-2019**: Monthly income appears fairly consistent.

   - **2020**: Significant spikes in August (6.22 x 10¹¹) and December, possibly reflecting special circumstances (maybe COVID-19-related budget adjustments).

   - **2021**: Again spikes in April and September but quite homogeneous across the year.

---

We can also have some quick insights of `Actual Amount` grouped by `Superior Agency`, `Agency` and `Managing Unit`

In [40]:
df.groupby('Superior Agency')['Actual Amount'].describe().T

Superior Agency  Advocacia-Geral da União  Controladoria-Geral da União  \
count                        3.732000e+03                  2.340000e+02   
mean                         7.509472e+04                  3.172387e+04   
std                          1.496635e+06                  1.252237e+05   
min                         -2.692618e+07                  2.000000e+00   
25%                          1.638625e+02                  1.678500e+02   
50%                          1.626260e+03                  1.131775e+03   
75%                          1.035065e+04                  9.887263e+03   
max                          6.382985e+07                  1.410733e+06   

Superior Agency  Ministério da Agricultura, Pecuária e Abastec  \
count                                             4.337900e+04   
mean                                              1.255170e+06   
std                                               2.695636e+07   
min                                              -5.000000e+07   
25%                                               5.190550e+02   
50%                                               8.944710e+03   
75%                                               6.334765e+04   
max                                               3.436513e+09   

Superior Agency  Ministério da Cidadania  \
count                       5.489000e+03   
mean                        6.590077e+05   
std                         1.625517e+07   
min                        -9.295862e+08   
25%                         2.926300e+02   
50%                         1.119333e+04   
75%                         1.070733e+05   
max                         4.162267e+08   

Superior Agency  Ministério da Ciência, Tecnologia, Inovações   \
count                                             3.102600e+04   
mean                                              1.089814e+06   
std                                               2.900641e+07   
min                                              -4.288879e+08   
25%                                               4.800000e+02   
50%                                               4.252595e+03   
75%                                               4.779214e+04   
max                                               3.208105e+09   

Superior Agency  Ministério da Defesa  Ministério da Economia  \
count                    9.021200e+04            1.885970e+05   
mean                     1.089369e+06            1.238519e+08   
std                      2.055599e+07            3.224854e+09   
min                     -1.526649e+08           -1.562859e+11   
25%                      3.260000e+02            1.402420e+03   
50%                      4.548860e+03            1.952564e+04   
75%                      5.711959e+04            6.050053e+05   
max                      2.649783e+09            7.711177e+11   

Superior Agency  Ministério da Educação  Ministério da Infraestrutura  \
count                      3.385350e+05                  3.554500e+04   
mean                       6.385683e+05                  3.655761e+06   
std                        5.805303e+07                  5.386872e+07   
min                       -1.160700e+08                 -1.280973e+08   
25%                        1.526500e+02                  1.347070e+03   
50%                        8.730000e+02                  2.437942e+04   
75%                        5.473115e+03                  3.821509e+05   
max                        1.903864e+10                  5.101423e+09   

Superior Agency  Ministério da Justiça e Segurança Pública  \
count                                         3.773800e+04   
mean                                          6.439728e+05   
std                                           1.000687e+07   
min                                          -7.243899e+08   
25%                                           6.768050e+02   
50%                                           1.214427e+04   
75%                                           8.765762e+04 

In [41]:
df.groupby('Agency')['Actual Amount'].describe().T

Agency  Advocacia-Geral da União - Unidades com vínculo direto  \
count                                        3.758000e+03        
mean                                         7.531658e+04        
std                                          1.491579e+06        
min                                         -2.692618e+07        
25%                                          1.643025e+02        
50%                                          1.629910e+03        
75%                                          1.036328e+04        
max                                          6.382985e+07        

Agency  Agência Espacial Brasileira  Agência Nacional de Aviação Civil  \
count                  8.200000e+01                       6.971000e+03   
mean                   3.178198e+05                       4.334492e+05   
std                    2.132394e+06                       4.365745e+06   
min                   -4.485200e+02                      -7.435685e+06   
25%                    2.031250e+01                       5.370950e+02   
50%                    8.681500e+01                       1.104995e+04   
75%                    1.562112e+03                       8.894059e+04   
max                    1.897520e+07                       2.195156e+08   

Agency  Agência Nacional de Energia Elétrica  Agência Nacional de Mineração  \
count                           4.137000e+03                   7.296000e+03   
mean                            1.460552e+07                   2.858518e+06   
std                             2.387396e+08                   3.365380e+07   
min                            -7.098133e+06                  -1.640522e+06   
25%                             8.472750e+03                   9.587653e+03   
50%                             8.242110e+04                   2.880510e+04   
75%                             7.494052e+05                   7.877431e+04   
max                             1.105000e+10                   1.098930e+09   

Agency  Agência Nacional de Saúde Suplementar  \
count                            6.786000e+03   
mean                             5.260659e+05   
std                              3.701703e+06   
min                             -1.966169e+06   
25%                              2.124485e+03   
50%                              4.810261e+04   
75%                              3.089984e+05   
max                              1.604105e+08   

Agency  Agência Nacional de Transportes Aquaviários  \
count                                  2.065000e+03   
mean                                   4.557652e+05   
std                                    9.551177e+06   
min                                   -2.733592e+05   
25%                                    5.807400e+02   
50%                                    3.146000e+03   
75%                                    1.452577e+04   
max                                    4.181163e+08   

Agency  Agência Nacional de Transportes Terrestres  \
count                                 6.395000e+03   
mean                                  1.451491e+06   
std                                   6.412120e+07   
min                                  -8.440152e+06   
25%                                   2.590205e+03   
50%                                   3.307544e+04   
75%                                   2.512209e+05   
max                                   5.101423e+09   

Agency  Agência Nacional de Vigilância Sanitária  Agência Nacional de Águas  \
count                               5.236000e+03               3.383000e+03   
mean                                6.841731e+05               7.049146e+05   
std                                 9.543350e+06               6.244800e+06   
min                                -7.594131e+07              -4.965438e+07   
25%                                 1.237050e+02               2.164550e+02   
50%                                 2.248062e+04               4.087570e+03   
75%                                 1.024704e+06

In [46]:
df.groupby('Managing Unit')['Actual Amount'].describe().T

Managing Unit  AG. NAC.DO PETROLEO GAS NAT.E BIOCOM  \
count                                  7.608000e+03   
mean                                   1.184038e+07   
std                                    5.858465e+08   
min                                   -1.175745e+10   
25%                                    1.797230e+03   
50%                                    1.800613e+04   
75%                                    8.922746e+04   
max                                    3.442000e+10   

Managing Unit  AGENCIA DE DESENVOLVIMENTO DA AMAZONIA  \
count                                        2.000000   
mean                                      1391.020000   
std                                        646.097608   
min                                        934.160000   
25%                                       1162.590000   
50%                                       1391.020000   
75%                                       1619.450000   
max                                       1847.880000   

Managing Unit  AGENCIA ESPACIAL BRASILEIRA/AEB  \
count                             8.200000e+01   
mean                              3.178198e+05   
std                               2.132394e+06   
min                              -4.485200e+02   
25%                               2.031250e+01   
50%                               8.681500e+01   
75%                               1.562112e+03   
max                               1.897520e+07   

Managing Unit  AGENCIA NACIONAL DE AVIACAO CIVIL  \
count                               6.972000e+03   
mean                                4.334118e+05   
std                                 4.365434e+06   
min                                -7.435685e+06   
25%                                 5.378050e+02   
50%                                 1.105379e+04   
75%                                 8.891470e+04   
max                                 2.195156e+08   

Managing Unit  AGENCIA NACIONAL DE AVIACAO CIVIL - F. AEROV.  \
count                                               4.000000   
mean                                            12210.435000   
std                                             10908.510647   
min                                              1331.540000   
25%                                              3768.530000   
50%                                             11953.635000   
75%                                             20395.540000   
max                                             23602.930000   

Managing Unit  AGENCIA NACIONAL DE ENERGIA ELETRICA  \
count                                  4.136000e+03   
mean                                   1.460854e+07   
std                                    2.387684e+08   
min                                   -7.098133e+06   
25%                                    8.448270e+03   
50%                                    8.219954e+04   
75%                                    7.497246e+05   
max                                    1.105000e+10   

Managing Unit  AGENCIA NACIONAL DE SAUDE SUPLEMENTAR  \
count                                   6.787000e+03   
mean                                    5.261112e+05   
std                                     3.701427e+06   
min                                    -1.966169e+06   
25%                                     2.126380e+03   
50%                                     4.813670e+04   
75%                                     3.092119e+05   
max                                     1.604105e+08   

Managing Unit  AGENCIA NACIONAL DE TELECOMUNICACOES-SEDE  \
count                                       9.922000e+03   
mean                                        1.561814e+06   
std                                         5.107585e+07   
min                                        -1.148723e+07   
25%                                         9.416000e+01   
50%                                         1.493330e+03   
75%                                         1.851195e+04   
max         

We can also group them by hierarchies:

In [51]:
df.groupby(['Superior Agency', 'Agency', 'Managing Unit'])['Actual Amount'].describe().T

Superior Agency                               Advocacia-Geral da União  \
Agency          Advocacia-Geral da União - Unidades com vínculo direto   
Managing Unit            COORD. GERAL DE ORC. FIN. E ANAL. CONT. - AGU   
count                                                 3.728000e+03       
mean                                                  7.514325e+04       
std                                                   1.497437e+06       
min                                                  -2.692618e+07       
25%                                                   1.630900e+02       
50%                                                   1.624320e+03       
75%                                                   1.035065e+04       
max                                                   6.382985e+07       

Superior Agency       Controladoria-Geral da União  \
Agency                Controladoria-Geral da União   
Managing Unit   DIRETORIA DE GESTAO INTERNA/SE/CGU   
count                                   158.000000   
mean                                  26935.733038   
std                                   93268.862455   
min                                       2.000000   
25%                                      60.287500   
50%                                     766.480000   
75%                                    5659.000000   
max                                  657603.790000   

Superior Agency                                                             \
Agency          Controladoria-Geral da União - Unidades com vínculo direto   
Managing Unit                             DIRETORIA DE GESTAO INTERNA(CGU)   
count                                                 7.500000e+01           
mean                                                  4.220010e+04           
std                                                   1.753453e+05           
min                                                   6.000000e+00           
25%                                                   4.117650e+02           
50%                                                   1.965700e+03           
75%                                                   1.752325e+04           
max                                                   1.410733e+06           

Superior Agency Ministério da Agricultura, Pecuária e Abastec  \
Agency                    Companhia Nacional de Abastecimento   
Managing Unit             COMPANHIA NACIONAL DE ABASTECIMENTO   
count                                            8.397000e+03   
mean                                             1.210700e+05   
std                                              1.843457e+06   
min                                             -6.539708e+06   
25%                                              1.958600e+02   
50%                                              1.943890e+03   
75%                                              1.402195e+04   
max                                              1.183330e+08   

Superior Agency                                              \
Agency          Empresa Brasileira de Pesquisa Agropecuária   
Managing Unit                              EMBRAPA-SETORIAL   
count                                          1.289400e+04   
mean                                           3.051815e+04   
std                                            3.008823e+05   
min                                           -4.667153e+06   
25%                                            1.345100e+02   
50%                                            1.536305e+03   
75%                                            8.240055e+03   
max                                            1.488925e+07   

Superior Agency                                               \
Agency                  Fundo de Defesa da Economia Cafeeira   
Managing Unit   FUNDO DE DEFESA DA ECON.CAFEEIRA/FUNCAFE/SPA   
count                                           3.869000e+03   
mean                                            8.816051e+06   
s